src/Test/Goblin/TH.hs

```diff
@@ -0,0 +1,192 @@
+{-# LANGUAGE TemplateHaskell       #-}
+module Test.Goblin.TH where
+
+import           Control.Monad (foldM, forM, unless)
+import           Data.Typeable (Typeable)
+import           Language.Haskell.TH
+import           TH.ReifySimple
+
+import Test.Goblin.Core
+
+
+--------------------------------------------------------------------------------
+-- Goblin instance derivation
+--------------------------------------------------------------------------------
+
+deriveGoblin :: Name -> Q [Dec]
+deriveGoblin name = do
+  (DataType _dName dTyVars _dCtx dCons) <- reifyDataType name
+  genomeVar <- VarT <$> newName "genome"
+  classParamNames <- forM dTyVars (const (newName "arg"))
+  let con = ensureSingleton dCons
+  ctx <- if null classParamNames
+            then (:[]) <$> [t| GeneOps $(pure genomeVar) |]
+            else wrapWithGoblinConstraints genomeVar classParamNames
+  decTy <- makeInstanceType genomeVar classParamNames
+  tink <- makeTinker con
+  conj <- makeConjure con
+  pure [InstanceD Nothing ctx decTy [tink, conj]]
+ where
+  ensureSingleton dCons =
+    case dCons of
+      []  -> error "deriveGoblin: cannot derive Goblin for a void type"
+      [x] -> x
+      _   -> error "deriveGoblin: cannot derive Goblin for a sum type"
```

Comment:
Any reason for this? We should be able to derive goblins for sum types...

Reply:
Hm. I had an explanation but as I wrote it out, realized it might not hold up. I'll have a think about this and reply more in depth.

Reply:
So let's look at an example sum-type, and an example `Goblin` instance for it:
```
data OneOrTwo a b = One a | Two a b
deriveAddShrinks ''OneOrTwo
deriveSeedGoblin ''OneOrTwo

instance ( Goblin g a, AddShrinks a
         , Goblin g b, AddShrinks b)
      => Goblin g (OneOrTwo a b) where

  tinker obj = do
    let tinks = [ do v <- conjure
                     let toIt x = case x of { One a -> a; Two _ _ -> v }
                     let val = toIt <$> obj
                     pure (One <$$> tinker val)
                , do (v1,v2) <- conjure
                     let toIt x = case x of { One _ -> (v1,v2); Two a b -> (a,b) }
                     let val = toIt <$> obj
                     pure (Two <$$> (tinker ((\(x,_) -> x) <$> val))
                               <**> (tinker ((\(_,x) -> x) <$> val)))
                ]
    i <- geneListIndex tinks
    ret <- tinks !! i
    tinkerRummagedOrConjureOrSave ret

  conjure = do
    let conjs = [ (One <$> conjure)
                , (Two <$> conjure <*> conjure)
                ]
    i <- geneListIndex conjs
    saveInBagOfTricks =<<
      (conjs !! i)
```

Since we only have "Functor access" to the `Gen` contents, in order to tinker with each branch we have to apply a partial function, but we can stitch two of them together by "covering" the other cases and conjuring up values.

Does that make sense to you, and does this seem like a reasonable way to construct instances? If so I can make the necessary changes to the TH derivation functions.

Reply:
If it's alright, I'm going to punt on this for the time being. We don't need to derive Goblin for any sum-types in `cardano-ledger-specs`. I have an idea how to generalize `deriveGoblins` for future cases, but I haven't had time to think through the ramifications fully.